AssemblyProvider.cs

```diff
@@ -0,0 +1,274 @@
+        pdb.Dispose();
+    }
+
+}
+
+public sealed class LinuxAssemblyProvider : AssemblyProvider {
+
+    private readonly ELF<ulong>         elf;
+    private readonly int                symOffset;
+    private readonly SymbolTable<ulong> symTable;
+
+    public LinuxAssemblyProvider(ELF<ulong> elf) {
+        this.elf = elf;
+
+        var text = (ProgBitsSection<ulong>)elf.GetSection(".text");
+        symOffset = (int)(text.Offset - text.LoadAddress);
+        symTable = (SymbolTable<ulong>)elf.GetSection(".symtab");
+    }
+
+    public OSPlatform Platform => OSPlatform.Linux;
+
+    public IEnumerable<Architecture> Architectures() {
+        yield return elf.Machine switch {
+            ElFMachine.AMD64 => Architecture.X64,
+            ElFMachine.AArch64 => Architecture.Arm64,
+            _ => throw new ArgumentException($"Unknown ELF architecture type. {Enum.GetName(typeof(ElFMachine), elf.Machine)}")
+        };
+    }
+
+    public Range FunctionFileRange(Architecture arch, string fName) {
+        Debug.Assert(Architectures().Contains(arch), "Tried to patch for unknown architecture.");
+
+        var func = symTable.Entries.Single(e => e.Name.Equals(fName));
+        var fnOffset = (int)func.Value + symOffset;
+        return new Range(fnOffset, fnOffset + (int)func.Size);
+    }
+
+    public void Dispose() {
+        elf.Dispose();
+    }
+
+}
+
+public sealed class MacosAssemblyProvider : AssemblyProvider {
+
+    private Dictionary<Architecture, MachO64> binaries;
+    private Signer?                           _signer;
+
+
+    public MacosAssemblyProvider(MachOFat fat) {
+        this.binaries = fat.Arches.ToDictionary(GetArchitecture, e => e.macho);
+        _signer = FindSignerProgram();
+    }
+
+    public OSPlatform Platform => OSPlatform.OSX;
+
+    public IEnumerable<Architecture> Architectures() {
+        return binaries.Keys;
+    }
+
+    public Range FunctionFileRange(Architecture arch, string fName) {
+        var macho = binaries[arch];
+
+        var sections = macho.LoadCommands.OfType<SegmentCommand64>().SelectMany(e => e.Sections).ToList();
+        var symTab = macho.LoadCommands.OfType<SymTabCommand>().Single();
+
+        var sym = symTab.Symbols.First(e => fName.Equals(e.Name));
+        var section = sections[(int)sym.SectionIndex!];
+        var funcSectionOffset = sym.Value - section.Address;
+        int fnOffset = (int)(funcSectionOffset + section.Offset + (ulong)macho.FileOffset);
+        return new Range(fnOffset, Index.End);
+    }
+
+    public void Dispose() {
+    }
+
+    private static Architecture GetArchitecture(MachOFatArch arch) {
+        return arch.macho.CpuType switch {
+            CpuType.X86_64 => Architecture.X64,
+            CpuType.Arm64 => Architecture.Arm64,
+            _ => throw new ArgumentException($"Unknown MachO architecture type. {Enum.GetName(typeof(CpuType), arch.macho.CpuType)}")
+        };
+    }
+    
+    
+    public bool PreCheck() {
```

Comment:
Can we just throw an exception from `FindSignerProgram` instead of having a separate `PreCheck`? It can be of another type which only shows message and not stack trace in the top level catch. Also means `_signer` is no longer nullable

Reply:
Done.